src/dust/dust.cpp

```diff
@@ -320,6 +457,344 @@ TaskStatus FluxSource(MeshData<Real> *md, const Real dt) {
+                         pmb->cellbounds.ncellsj(IndexDomain::entire),
+                         pmb->cellbounds.ncellsi(IndexDomain::entire));
+
+  int maxCalls = 0, maxSize = 1;
+  auto &dfloor = dust_pkg->template Param<Real>("dfloor");
+
+  int maxSize0 = 1;
+  Real massd = 0.0;
+  Kokkos::parallel_reduce(
+      "coag::maxSize0",
+      Kokkos::MDRangePolicy<Kokkos::Rank<4>>(
+          {0, kb.s, jb.s, ib.s}, {md->NumBlocks(), kb.e + 1, jb.e + 1, ib.e + 1}),
+      KOKKOS_LAMBDA(const int b, const int k, const int j, const int i, Real &lsum,
+                    int &lmax) {
+        geometry::Coords<GEOM> coords(vmesh.GetCoordinates(b), k, j, i);
+        for (int n = 0; n < nspecies; ++n) {
+          Real &dens_d = vmesh(b, dust::cons::density(n), k, j, i);
+          lsum += dens_d * coords.Volume();
+        }
+        for (int n = nspecies - 1; n >= 0; --n) {
+          Real &dens_d = vmesh(b, dust::cons::density(n), k, j, i);
+          if (dens_d > dfloor) {
+            lmax = std::max(lmax, n);
+            break;
+          }
+        }
+      },
+      massd, Kokkos::Max<int>(maxSize0));
+#ifdef MPI_PARALLEL
+  // Sum the perturbations over all processors
+  MPI_Reduce(MPI_IN_PLACE, &maxSize0, 1, MPI_INT, MPI_MAX, 0, MPI_COMM_WORLD);
+  MPI_Reduce(MPI_IN_PLACE, &massd, 1, MPI_PARTHENON_REAL, MPI_SUM, 0, MPI_COMM_WORLD);
+#endif // MPI_PARALLEL
+
+  Real sumd1 = 0.0;
+#ifdef COAG_DEBUG
+  ParArray3D<Real> floor_tmp("floor_tmp", 17, nspecies, 2);
+#endif
+  for (int b = 0; b < md->NumBlocks(); b++) {
+    parthenon::par_for_outer(
+        DEFAULT_OUTER_LOOP_PATTERN, "Dust::Coagulation", parthenon::DevExecSpace(),
+        scr_size, scr_level, kb.s, kb.e, jb.s, jb.e, ib.s, ib.e,
+        KOKKOS_LAMBDA(parthenon::team_mbr_t mbr, const int k, const int j, const int i) {
+          // code-to-physical unit
+          // one-cell coagulation
+          const int nDust_live = (vmesh.GetUpperBound(b, dust::prim::density()) -
+                                  vmesh.GetLowerBound(b, dust::prim::density()) + 1);
+
+          const Real dens_g = vmesh(b, gas::prim::density(0), k, j, i) * den0;
+          Real dt_sync = dt * time0;
+          const Real time1 = time * time0;
+
+          geometry::Coords<GEOM> coords(vmesh.GetCoordinates(b), k, j, i);
+          const auto &hx = coords.GetScaleFactors();
+          Real rad = hx[2];
+          if constexpr (GEOM == Coordinates::cylindrical) {
+            rad = hx[1];
+          }
+          const Real Omega_k = 1.0 / std::sqrt(rad) / rad; // code unit
+          // const Real vol1 = coords.Volume() * vol0;
+
+          int nCall1 = 0;
+
+          const Real pres = vmesh(b, gas::prim::pressure(0), k, j, i);
+          const Real cs = std::sqrt(gamma * pres / dens_g * den0) * vel0;
+          const Real omega1 = Omega_k / time0;
+          const int nm = nspecies;
+
+          ScratchPad1D<Real> rhod(mbr.team_scratch(scr_level), nspecies);
```

Comment:
Maybe we should think a bit more about how to store this information. I could see this being okay for when you have > 100 dust species, but maybe not being very performant for < 20 species. Maybe in those smaller cases, you'd want a 2D scratchpad that is `(nspecies,nx1)` large.

Reply:
For coagulation run, nspecies is usually larger than 100.